### DISCUSSION TITLE
Thinking about a Roadmap

### DISCUSSION THREAD

**Discussion by dcorson-ticino-com** [jamulussoftware/jamulus#1846] [Ideas]
Now that 3_8_0 is out the door I want to bring up the roadmap again.  I thought we had talked about this at one point, but I can now find nothing.  I think it would be very worth while to have an mid-term idea of what we want to do, how Jamulus should evolve, to organise best use of our energy for development.  This is also closely related with how we think the end of the pandemic caused restrictions will effect the use of Jamulus.

My personal impression how the lifting of the pandemic restrictions will effect the use of Jamulus is that Jamulus will be less used by amateur groups, for local choirs, for example, but will continue to be used by musicians who value the possiblities to reduce travel.  My reasoning being that in my experience choirs and amateur orchestras are at least 50% social events for the participants and because of that they will return to in-person events as soon as possible (as has already the orchestra I play with).  On the other hand musicians wanting to rehearse or jam are glad to have the possibility to play together without traveling, will be looking for quality sound and may be willing to invest a little for that.

I see the following key aspects for the future from the discussions on the forums (in no particular order):

1) Inclusion of new Hardware
      Tablets, Android, IOS
2) Master mix
      Another controls your mix
3) Multi channel fader
      Separate L & R channels to two mono/stereo channels each with their own faders and pan
4) Single window UI
      Mainly for tablets?
5) Better audio quality
      Appears to be the main reason why Sonobus is being used
6) Windows drivers
      Eliminate ASIO4ALL
7) Updates of programming libraries, Qt, rewrite with API to separate UI from workers. 
8) Session Management
      Authentication, user in- and exclusion 
9) IPv6
10) Favorites
      ?Bookmarks? for me it is done and OK to include

My thoughts on the key aspects in light of that impression is as follows:

1) New Hardware - I am sceptical, lots of work for devices that probably will only be connected by wifi/GSM lowering the quality for everyone. Only if a developer with hardware knowledge decides to take this and run with it 'til it really works without loosing quality.
2) Master mix - if we do this, do it quickly now before the choirs go back to rehearsing in-person.
3) Multi-channel fader - yes, is certainly useful.
4) Single window UI - this is mostly interesting for tablets, so if we don't do Android and IOS not needed.  On the other hand if wanted I could do this as I now know how from the NewUI and it is relatively simple using Qt docking windows with the existing window classes.  If this is seen as being a more modern interface on a PC, why not?
5) Better audio quality - a good addition.  My question: for clients doing no compression could we eliminate one stage of buffering and reduce the latency at the same time?
6) Windows drivers - I am not sure what to think, PortAudio does not seem to offer the advantages (that at least I) imagined. Is to support one of the Win10 native interfaces a solution? Or do we keep the status quo assuming the users of the future will be willing to invest a little for better quality with an interface.
7) Rewrite - Mamouth job, rewriting 10 years (or more) of development. Will need to be in the background.
8) Session Management - A very welcome addition, but it looks like a lot of work.
9) IPv6 - Seems to be a very big bite, lots of work.  Is an incremental implementation possible and/or useful?

IMHO Jamulus has reached a level of maturity that is very good and is amaisingly bug free.  
My suggestion is that we take one of the key aspects as the main feature for each of the coming  major versions, starting with Master Mix.  While I think that many would love the Multi-channel fader I expect almost everyone with a good connection would like to better their audio quality and would suggest that as the following step.

Thanks for following me this far.  Have I missed key features we can't live without?  What are your thoughts?  
There is a lot we could do, especially with the continuing bug fixes etc., let's put some order in it, realizing, of course, that none of us can work on this without limit.

**Comment by gene96817** [comment]
Re: #1:  We should be careful about putting effort into devices that cannot use ethernet or if the screen is very small. I think we should strive for a common user experience across all platforms. Android and iOS are good platforms if they are tablets. I would be concerned about how much the UI has to change to fit a smartphone screen.... Around here, the schools have made a big decision to use Chromebooks. In a few months, Chromebooks will ship with Crostini (supposed to support Linux apps). It would be very interesting to make Jamulus available for the music departments.

Re: #6: "Fighting" with Microsoft's choices of drivers has always been painful. Do we believe Microsoft will soon be shipping with Windows using Linux under the UI? That could change the drivers. On the other hand, it would be consistent with Microsoft's past to keep the existing driver architecture.

Re: #9: Packaging up all the existing IPv6 work into an incremental step would be very interesting. The Jamulus community is large enough to make a very good read on the reality of IPv6. We could be surprised that IPv6 delivers an across-the-board reduction in latency. (There are a lot of complicated tricks on the Internet to keep IPv4 working. That has to add latency.)

**Comment by WolfganP** [comment]
Great discussion. I would add:
3- Expand to true multichannel upstream from the client, for better support of -better than stereo- interfaces (ie multichannel mixing boards)
11- Instrumentation: have detailed technical stats about buffers, queues, timers, cpu/threads, lost packets, ... so better diagnostic of issues and help to other users can be delivered.
12- Server side multi-channels' raw streaming (to allow better mixing via a DAW of choice, OBS, or some other audio chain of choice)

**Comment by chrisrimple** [comment]
As a person involved in the choral world, I agree that choral groups and choirs are unlikely to continue using Jamulus after the pandemic allows them to rehearse in person.  Where I do think Jamulus could continue to see a lot of use, and how I will be using it, is for remote rehearsals with my local band.  Pre-Covid, we were rehearsing as a group in person once per week; during Covid, we've been rehearsing twice per week on Jamulus.  Post-Covid, we're talking about returning to in person rehearsals once per week, but also continuing with Jamulus rehearsals once per week.  Having Jamulus available makes it easier for us to rehearse without the need to haul equipment, etc.

From the perspective of a band musician, most of the possible future features have no relevance.  Instead, reducing latency, eliminating server sound dropouts, etc. are more important.  Also improving sound quality and solving for Windows drivers, making it easier for Windows-based musicians to get started with Jamulus.
